Various fixes to the allocator and the allocator tests.

hlef opened the pull request:
- Demonstrate a race in the allocator between claims and ephemeral claims, and how this race may lead the allocator to ignore a claim and make quota disappear into the void. Fix this by checking the number of claims on an object when removing it from the hazard quarantine (surprisingly small patch, hopefully I got it right!). This corner-case was reported by Reto Achermann (UBC) as we are working on formally verifying the allocator.
- Improve allocator tests to make them more readable and expressive.
- The tests have multiple instances of using `is_valid` to check tags when they should be using something not subject to CSE (here be dragons). I made the mistake myself when updating the tests. Address this by adding a C++ wrapper `is_valid_temporal` for `__builtin_cheri_tag_get_temporal` and using it where appropriate in the tests. Make sure the behavior w.r.t. CSE is well-documented.

Entirely artisanal, no AI involved. All hallucinations are mine.

hlef committed:
allocator-tests: improve hazard tests.

Waiting on `heap_quota_remaining(MALLOC_CAPABILITY) < MALLOC_QUOTA` to
check if the lambda was freed is super confusing (especially since the
test does not document why it is doing this). Instead, wait until the
allocator returns `-EINVAL` for the pointers we freed, which indicates
that 1) the ephemeral claims have been released and 2) the lambda was
freed since it called the allocator.

This also has the advantage of documenting the difference between
`heap_free` returning `-EPERM` (the chunk has been freed but is waiting
on the hazard list) and `-EINVAL` (the chunk has actually been freed).

Signed-off-by: Hugo Lefeuvre <hugo.lefeuvre@ubc.ca>

hlef committed:
allocator-test: test races between ephemeral claims and claims.

This expands the current hazard tests to include a situation where:

T1: allocate `ptr` with heap quota 1
T2: ephemeral claim on `ptr`
T1: free `ptr` with heap quota 1
T1: claim `ptr` with heap quota 2
T2: release ephemeral claim on `ptr`

In such a case, quota 2 should have a claim on `ptr` and thus the
pointer should not be freed. However this test failed prior to
59a7783e01a7f5b44417b6beee64c834a72fae62 as the allocator ignored quota
2's claim and freed `ptr`.

Signed-off-by: Hugo Lefeuvre <hugo.lefeuvre@ubc.ca>

hlef committed:
cheri.hh: add `is_valid_temporal` wrapper.

Add `is_valid_temporal`, a C++ wrapper for
`__builtin_cheri_tag_get_temporal`, and clarify the behavior of
`is_valid` and `is_valid_temporal` w.r.t. CSE.

Signed-off-by: Hugo Lefeuvre <hugo.lefeuvre@ubc.ca>

hlef committed:
allocator-test: use `is_valid_temporal()` where CSE might bite us

Three places in the allocator tests call `is_valid` multiple times on
the same capability and expect the tag to be cleared in between. CSE
might be a problem there, so use the temporal variant.

Signed-off-by: Hugo Lefeuvre <hugo.lefeuvre@ubc.ca>

hlef committed:
alloc.h: fix race between claim and ephemeral claim.

There is no check for the number of claims in `mspace_free`, which
causes issues when an object is claimed while in the hazard quarantine.
As soon as it leaves the hazard quarantine, the allocator assumes that
there are no outstanding claims, and proceeds painting and freeing
`ptr`. This may happen if an object is claimed while an ephemeral claim
is holding the allocator from freeing it.

This bug is doubly (triply?) bad: we free an object we said was claimed,
we do not refund the quota associated to this claim, and we leak any
associated claim record object(s), since there would, after free, be no
initial pointer to the list thereof.

This corner case was reported by Reto Achermann (UBC) as part of an
ongoing verification effort of the memory allocator.

Signed-off-by: Hugo Lefeuvre <hugo.lefeuvre@ubc.ca>

nwf reviewed (changes requested):
Terrific work and great find!  Very minor nits.

I wonder if it's worth expanding the allocator _fuzz_ test to...
1. Use some async() jobs to take ephemeral claims...
2. Take full claims instead of allocations, some fraction of the time, during operation (multiply-listing claimed objects in `allocations`)
3. Ensure that any decrease in is quota is accounted for by the allocated objects in its collection (the sum of allocated sizes may appear larger than the decrease in quota, due to claims, but a quota leak would, WHP, eventually be revealed).

hlef commented on tests/allocator-test.cc:
```diff
@@ -644,11 +649,19 @@ namespace
+		       (claimSize <= allocSize + CHERIOTHeapMinChunkSize),
+		     "{}-byte allocation claimed as {} bytes",
+		     allocSize,
+		     claimSize);
```
Also, this test will be annoying if we want to use the test suite on the formally verified Rust implementation at some point.

hlef committed:
allocator-test: remove test for internal implementation detail.

We currently test if the claim size returned by `heap_claim()` is
constrained between the allocation size and allocation size plus the
size of a claim object.

As mentioned by @nwf we do not account for the size of a claim object in
the return value of `heap_claim()`, so `(claimSize <= allocSize +
CHERIOTHeapMinChunkSize)` should always be true.

In addition, CHERIOTHeapMinChunkSize is an implementation detail that
may not be provided by other allocator implementations, such as the
formally-verified Rust allocator we are developing at UBC. This will be
a problem if we want to re-use the test suite.

Signed-off-by: Hugo Lefeuvre <hugo.lefeuvre@ubc.ca>

hlef commented:
@nwf All comments should be fixed now!

nwf reviewed (approved):
LGTM!
